greenplumpython/gpapply.py

```diff
@@ -13,94 +15,59 @@ def randomStringType(stringLength=10):
-    
     return select_func
 
     drop_func = "DROP TYPE " + typeName + " CASCADE;"
-    
-def createTypeFunc(output, typeName):
-    typeSQL = output[0] + " " + output[1]
-    for i in range(2, len(output)):
-        if (i%2 == 0):
-            typeSQL = typeSQL + ",\n" + output[i] + " " + output[i+1]
-        i = i + 1
+
+def createTypeFunc(sig, typeName):
+    typeSQL = ""
+    for i  in range(0,len(sig)):
+        if i == 0:
+            for j, col in enumerate(sig[i]):
+                typeSQL += col + " " + sig[i][col]
+        else:
+            for j, col in enumerate(sig[i]):
+                typeSQL += ",\n" + col + " " + sig[i][col]
     typeSQL = "CREATE TYPE " + typeName + " AS (\n" + typeSQL + "\n);"
     return typeSQL
 
+def gpApply(dataframe, py_func, db, output, clear_existing = True, runtime_id = 'plc_python', runtime_type = 'plcontainer', **input_py_func_extra_args):
 
-def pythonApply(input, output, index, py_func, gpdb_tbl_name, output_tbl_name):
     s = inspect.getsource(py_func)
     #gpdb_tbl_name = "testtbl"
     function_name = randomString()
     typeName = randomStringType()
-    print(function_name)
     params = []
     columns = []
-    i = 0
-    while i < len(input):
-        if input[i] == index:
-            params.append(input[i]+" "+input[i+1])
-        else:
-            params.append(input[i]+" "+input[i+1]+"[]")
-        columns.append(input[i])
-        i += 2
+    for i, col in enumerate(dataframe.table_metadata.signature):
+        for j, column in enumerate(col):
+            params.append(column+" "+col[column] + "[]")
+            columns.append(column)
         
-    create_type_sql = createTypeFunc(output, typeName)
+    create_type_sql = createTypeFunc(output.signature, typeName)
     function_body = "CREATE OR REPLACE FUNCTION %s(%s) RETURNS %s AS $$\n %s return %s(%s) $$ LANGUAGE plpythonu;" % (function_name,",".join(params),typeName,s,py_func.__name__,",".join(columns))
-
-    select_sql = pythonExec(input, index, gpdb_tbl_name, function_name, typeName, output_tbl_name)
-
+    select_sql = pythonExec(dataframe, function_name, typeName, output.name, input_py_func_extra_args)
     drop_sql = "DROP TYPE " + typeName + " CASCADE;"
-    
-    
-    
-    conn = psycopg2.connect(database="test", user="gpadmin", port="15432")
-    cur = conn.cursor()
-    cur.execute("drop table if exists weather_output;")
-    conn.commit()
-    cur.execute(create_type_sql)
-    conn.commit()
-
-    cur.execute(function_body)
-    conn.commit()
-    
-    
-    cur.execute(select_sql)
-    
-    cur.execute(drop_sql)
-    output_select_sql = "select * from " + output_tbl_name + ";"
-    cur.execute(output_select_sql)
-    res = cur.fetchall()
-    conn.commit()
-    return res
+    if clear_existing:
+        drop_table_sql = "drop table if exists %s;" % output.name
+        db.execute(drop_table_sql)
+    db.execute(create_type_sql)
```

Comment:
It's better to make these sql execution in transaction block.

Reply:
we already have a story for this #172082257